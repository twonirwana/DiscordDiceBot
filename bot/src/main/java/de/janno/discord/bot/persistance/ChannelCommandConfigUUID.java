package de.janno.discord.bot.persistance;

import lombok.NonNull;
import lombok.Value;

import java.time.LocalDateTime;
import java.util.UUID;

@Value
public class ChannelCommandConfigUUID {
    @NonNull
    String command;
    @NonNull
    LocalDateTime creationDate;
    @NonNull
    UUID configUUID;

}
