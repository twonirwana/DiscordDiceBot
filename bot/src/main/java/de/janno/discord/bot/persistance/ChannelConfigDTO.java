package de.janno.discord.bot.persistance;

import lombok.AllArgsConstructor;
import lombok.NonNull;
import lombok.Value;

import java.util.UUID;

@Value
@AllArgsConstructor
public class ChannelConfigDTO {

    @NonNull
    UUID configUUID;
    Long guildId;
    long channelId;
    Long userId;
    @NonNull
    String commandId;
    @NonNull
    String configClassId;
    @NonNull
    String config;
}
