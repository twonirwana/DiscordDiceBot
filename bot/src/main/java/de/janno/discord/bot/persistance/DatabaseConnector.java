package de.janno.discord.bot.persistance;

import com.google.common.collect.ImmutableSet;
import io.micrometer.core.instrument.binder.db.DatabaseTableMetrics;
import lombok.NonNull;
import org.h2.jdbcx.JdbcConnectionPool;
import org.jetbrains.annotations.Nullable;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.SQLException;

import static io.micrometer.core.instrument.Metrics.globalRegistry;

public class DatabaseConnector {

    private final JdbcConnectionPool connectionPool;

    public DatabaseConnector(@NonNull String url, @Nullable String user, @Nullable String password) {
        connectionPool = JdbcConnectionPool.create(url, user, password);
        new DatabaseTableMetrics(connectionPool, "h2", "MESSAGE_DATA", ImmutableSet.of()).bindTo(globalRegistry);
        new DatabaseTableMetrics(connectionPool, "h2", "MESSAGE_CONFIG", ImmutableSet.of()).bindTo(globalRegistry);
        new DatabaseTableMetrics(connectionPool, "h2", "CHANNEL_CONFIG", ImmutableSet.of()).bindTo(globalRegistry);

    }

    public Connection getConnection() throws SQLException {
        return connectionPool.getConnection();
    }

    public void dispose() {
        connectionPool.dispose();
    }

    public DataSource getDataSource(){
        return connectionPool;
    }
}
