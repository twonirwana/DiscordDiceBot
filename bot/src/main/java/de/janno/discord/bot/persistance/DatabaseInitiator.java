package de.janno.discord.bot.persistance;

import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.io.Resources;
import io.micrometer.core.instrument.util.IOUtils;
import lombok.NonNull;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;
import org.h2.jdbcx.JdbcConnectionPool;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.sql.*;
import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
public class DatabaseInitiator {

    private final static List<String> MIGRATION_FILES = ImmutableList.<String>builder()
            .add("1_base.sql")
            .add("2_configTable.sql")
            .add("3_channelConfigUser.sql")
            .add("4_aliasCommandIdCleanUp.sql")
            .add("5_channelConfigFix.sql")
            .build();

    public static void initialize(@NonNull String url, @Nullable String user, @Nullable String password) {
        JdbcConnectionPool connectionPool = JdbcConnectionPool.create(url, user, password);

        List<Migration> allMigrations = readMigrations();
        log.info("found migrations: {}", allMigrations.stream().map(Migration::getName).collect(Collectors.joining(", ")));
        List<String> alreadyAppliedMigrations = getAlreadyAppliedMigrations(connectionPool);
        log.info("db migration status: {}", alreadyAppliedMigrations);
        applyMissingMigrations(allMigrations, alreadyAppliedMigrations, connectionPool);
    }

    private static void applyMissingMigrations(List<Migration> allMigrations, List<String> alreadyAppliedMigrations, JdbcConnectionPool connectionPool) {
        Preconditions.checkArgument(allMigrations.stream().map(Migration::getOrder).distinct().count() == allMigrations.size(), "Duplicate migration order");
        Preconditions.checkArgument(allMigrations.stream().map(Migration::getName).distinct().count() == allMigrations.size(), "Duplicate migration name");

        allMigrations.stream()
                .filter(m -> !alreadyAppliedMigrations.contains(m.getName()))
                .sorted(Comparator.comparing(Migration::getOrder))
                .forEach(m -> {
                    log.info("Start executing {}", m.getName());
                    try (Connection connection = connectionPool.getConnection()) {
                        Statement statement = connection.createStatement();
                        statement.execute(m.getSql());
                        try (PreparedStatement preparedStatement =
                                     connection.prepareStatement("INSERT INTO DB_VERSION(MIGRATION_NAME, CREATION_DATE) VALUES (?, ?)")) {
                            preparedStatement.setString(1, m.getName());
                            preparedStatement.setTimestamp(2, Timestamp.valueOf(LocalDateTime.now()));
                            preparedStatement.execute();
                        }
                    } catch (SQLException e) {
                        log.error("Error in {}", m.getName());
                        throw new RuntimeException(e);
                    }
                    log.info("Finish executing {}", m.getName());
                });
    }

    private static List<String> getAlreadyAppliedMigrations(JdbcConnectionPool connectionPool) {
        try (Connection connection = connectionPool.getConnection()) {
            Statement statement = connection.createStatement();
            ResultSet resultSet = statement.executeQuery("select MIGRATION_NAME from DB_VERSION");
            ImmutableList.Builder<String> builder = ImmutableList.builder();
            while (resultSet.next()) {
                builder.add(resultSet.getString("MIGRATION_NAME"));
            }
            return builder.build();
        } catch (SQLException e) {
            return List.of();
        }
    }

    private static List<Migration> readMigrations() {
        return MIGRATION_FILES.stream()
                .map(DatabaseInitiator::readMigration)
                .toList();
    }

    private static Migration readMigration(final String fileName) {
        Preconditions.checkArgument(fileName.contains("_"), "Wrong file format: {}", fileName);
        Preconditions.checkArgument(fileName.endsWith("sql"), "Wrong file format: {}", fileName);
        String content;
        try {
            content = IOUtils.toString(Resources.getResource("db/migrations/" + fileName).openStream(), StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        int firstUnderscoreIndex = fileName.indexOf("_");
        String order = fileName.substring(0, firstUnderscoreIndex);
        String name = fileName.substring(firstUnderscoreIndex + 1, fileName.length() - 4);
        return new Migration(order, name, content);
    }


    @Value
    private static class Migration {
        String order;
        String name;
        String sql;
    }
}
