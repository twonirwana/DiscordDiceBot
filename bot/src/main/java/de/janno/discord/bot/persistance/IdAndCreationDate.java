package de.janno.discord.bot.persistance;

import lombok.Value;

import java.time.LocalDateTime;

@Value
public class IdAndCreationDate {
        long id;
        LocalDateTime creationDate;
}
