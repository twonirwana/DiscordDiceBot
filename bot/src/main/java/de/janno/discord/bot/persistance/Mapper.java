package de.janno.discord.bot.persistance;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import lombok.NonNull;


public class Mapper {

    public static final String NO_PERSISTED_STATE = "None";
    private static final ObjectMapper mapper = new ObjectMapper(new YAMLFactory());

    public static String serializedObject(@NonNull Object object) {
        try {
            return mapper.writeValueAsString(object);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException(e);
        }
    }

    public static <T> T deserializeObject(@NonNull String object, @NonNull Class<T> classOfObject) {
        try {
            return mapper.readValue(object, classOfObject);
        } catch (JsonProcessingException e) {
            throw new IllegalStateException(e);
        }
    }

}
