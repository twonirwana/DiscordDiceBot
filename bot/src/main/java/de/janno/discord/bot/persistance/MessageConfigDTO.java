package de.janno.discord.bot.persistance;

import lombok.AllArgsConstructor;
import lombok.NonNull;
import lombok.Value;
import javax.annotation.Nullable;

import java.util.UUID;

@Value
@AllArgsConstructor
public class MessageConfigDTO {

    @NonNull
    UUID configUUID;
    @Nullable
    Long guildId;
    long channelId;
    @NonNull
    String commandId;
    @NonNull
    String configClassId;
    @NonNull
    String config;

}
