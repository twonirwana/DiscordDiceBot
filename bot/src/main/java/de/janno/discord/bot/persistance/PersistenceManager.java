package de.janno.discord.bot.persistance;

import lombok.NonNull;

import java.util.Optional;
import java.util.Set;
import java.util.UUID;

public interface PersistenceManager {

    @NonNull Optional<MessageConfigDTO> getMessageConfig(@NonNull UUID configUUID);

    @NonNull Optional<MessageConfigDTO> getConfigFromMessage(long channelId, long messageId);

    void saveMessageConfig(@NonNull MessageConfigDTO messageConfigDTO);

    @NonNull Optional<MessageDataDTO> getMessageData(long channelId, long messageId);

    @NonNull Set<Long> getAllMessageIdsForConfig(@NonNull UUID configUUID);

    void deleteStateForMessage(long channelId, long messageId);

    @NonNull Set<Long> deleteMessageDataForChannel(long channelId);

    void saveMessageData(@NonNull MessageDataDTO messageState);

    Set<Long> getAllGuildIds();

    @NonNull Optional<ChannelConfigDTO> getChannelConfig(long channelId, String configClassId);

    void saveChannelConfig(@NonNull ChannelConfigDTO channelConfigDTO);

    void deleteChannelConfig(long channelId, String configClassId);
}
