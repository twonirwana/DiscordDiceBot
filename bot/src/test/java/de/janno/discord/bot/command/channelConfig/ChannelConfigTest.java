package de.janno.discord.bot.command.channelConfig;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import de.janno.discord.bot.persistance.PersistenceManager;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import static org.mockito.Mockito.mock;

@ExtendWith(SnapshotExtension.class)
public class ChannelConfigTest {
    ChannelConfigCommand underTest;
    private Expect expect;


    @BeforeEach
    void setup() {
        underTest = new ChannelConfigCommand(mock(PersistenceManager.class));
    }

    @Test
    public void getCommandDefinition() {
        expect.toMatchSnapshot(underTest.getCommandDefinition());
    }

    @Test
    public void getId() {
        expect.toMatchSnapshot(underTest.getCommandId());
    }
}
