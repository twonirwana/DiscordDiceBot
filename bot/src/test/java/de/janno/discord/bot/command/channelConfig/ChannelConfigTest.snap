de.janno.discord.bot.command.channelConfig.ChannelConfigTest.getCommandDefinition=[
  {
    "name": "channel_config",
    "description": "Configure options in this channel",
    "nameLocales": [ ],
    "descriptionLocales": [
      {
        "locale": "de",
        "description": "Konfiguration des Bots in diesem Kanal"
      },
      {
        "locale": "pt_BR",
        "description": "Configure as opções neste canal"
      }
    ],
    "options": [
      {
        "type": "SUB_COMMAND",
        "name": "save_direct_roll_config",
        "nameLocales": [
          {
            "locale": "de",
            "name": "speicher_direkt_wurf_konfig"
          },
          {
            "locale": "pt_BR",
            "name": "salvar_config_rolagem_direta"
          }
        ],
        "description": "add or update the channel config",
        "descriptionLocales": [
          {
            "locale": "de",
            "description": "Hinzufügen und aktualisieren von Kanalkonfiguration"
          },
          {
            "locale": "pt_BR",
            "description": "adiciona ou atualiza a configuração do canal"
          }
        ],
        "required": false,
        "choices": [ ],
        "options": [
          {
            "type": "STRING",
            "name": "answer_format",
            "nameLocales": [
              {
                "locale": "de",
                "name": "antwort_format"
              },
              {
                "locale": "pt_BR",
                "name": "formato_resultado"
              }
            ],
            "description": "How the answer will be displayed",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Wie die Antwort formatiert wird"
              },
              {
                "locale": "pt_BR",
                "description": "Como o resultado será exibido"
              }
            ],
            "required": false,
            "choices": [
              {
                "name": "full",
                "value": "full",
                "nameLocales": [
                  {
                    "locale": "de",
                    "choice": "voll"
                  },
                  {
                    "locale": "pt_BR",
                    "choice": "completo"
                  }
                ]
              },
              {
                "name": "without_expression",
                "value": "without_expression",
                "nameLocales": [
                  {
                    "locale": "de",
                    "choice": "ohne_ausdruck"
                  },
                  {
                    "locale": "pt_BR",
                    "choice": "sem_formula"
                  }
                ]
              },
              {
                "name": "only_dice",
                "value": "only_dice",
                "nameLocales": [
                  {
                    "locale": "de",
                    "choice": "nur_würfel"
                  },
                  {
                    "locale": "pt_BR",
                    "choice": "apenas_dados"
                  }
                ]
              },
              {
                "name": "compact",
                "value": "compact",
                "nameLocales": [
                  {
                    "locale": "de",
                    "choice": "kompakt"
                  },
                  {
                    "locale": "pt_BR",
                    "choice": "compacto"
                  }
                ]
              },
              {
                "name": "minimal",
                "value": "minimal",
                "nameLocales": [
                  {
                    "locale": "pt_BR",
                    "choice": "minimo"
                  }
                ]
              }
            ],
            "options": [ ],
            "autoComplete": false
          },
          {
            "type": "STRING",
            "name": "dice_image_style",
            "nameLocales": [
              {
                "locale": "de",
                "name": "würfel_bild_stil"
              },
              {
                "locale": "pt_BR",
                "name": "estilo_imagem_dados"
              }
            ],
            "description": "If and in what style the dice throw should be shown as image",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Ob und in welchen Stil Würfelbilder gezeigt werden"
              },
              {
                "locale": "pt_BR",
                "description": "Se os dados devem ser mostrados como imagens e em qual estilo"
              }
            ],
            "required": false,
            "choices": [
              {
                "name": "none",
                "value": "none",
                "nameLocales": [
                  {
                    "locale": "pt_BR",
                    "choice": "nenhum"
                  }
                ]
              },
              {
                "name": "polyhedral_3d",
                "value": "polyhedral_3d",
                "nameLocales": [ ]
              },
              {
                "name": "polyhedral_alies_v2",
                "value": "polyhedral_alies_v2",
                "nameLocales": [ ]
              },
              {
                "name": "polyhedral_knots",
                "value": "polyhedral_knots",
                "nameLocales": [ ]
              },
              {
                "name": "polyhedral_RdD",
                "value": "polyhedral_RdD",
                "nameLocales": [ ]
              },
              {
                "name": "fate",
                "value": "fate",
                "nameLocales": [ ]
              },
              {
                "name": "d6_dots",
                "value": "d6_dots",
                "nameLocales": [ ]
              },
              {
                "name": "polyhedral_2d",
                "value": "polyhedral_2d",
                "nameLocales": [ ]
              },
              {
                "name": "polyhedral_alies_v1",
                "value": "polyhedral_alies_v1",
                "nameLocales": [ ]
              }
            ],
            "options": [ ],
            "autoComplete": false
          },
          {
            "type": "STRING",
            "name": "dice_image_color",
            "nameLocales": [
              {
                "locale": "de",
                "name": "würfel_bild_farbe"
              },
              {
                "locale": "pt_BR",
                "name": "cor_dos_dados"
              }
            ],
            "description": "The default color option. Can be influenced by the `col` operator",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Die Standardfarbe, kann mit noch mit `col` beeinflusst werden."
              },
              {
                "locale": "pt_BR",
                "description": "A opção de cor padrão. Pode ser influenciada pelo operador `col`"
              }
            ],
            "required": false,
            "choices": [ ],
            "options": [ ],
            "autoComplete": true
          },
          {
            "type": "BOOLEAN",
            "name": "always_sum_result",
            "nameLocales": [
              {
                "locale": "de",
                "name": "ergebnis_zusammen_zählen"
              },
              {
                "locale": "pt_BR",
                "name": "sempre_somar_resultado"
              }
            ],
            "description": "Always sum the results of the dice expressions",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Addiere die Würfel immer zusammen"
              },
              {
                "locale": "pt_BR",
                "description": "Sempre soma os resultados das fórmulas de dados"
              }
            ],
            "required": false,
            "choices": [ ],
            "options": [ ],
            "autoComplete": false
          }
        ],
        "autoComplete": false
      },
      {
        "type": "SUB_COMMAND_GROUP",
        "name": "delete_direct_roll_config",
        "nameLocales": [
          {
            "locale": "de",
            "name": "lösch_direkt_wurf_konfig"
          },
          {
            "locale": "pt_BR",
            "name": "deletar_config_rolagem_direta"
          }
        ],
        "description": "remove the current channel config",
        "descriptionLocales": [
          {
            "locale": "de",
            "description": "Entferne die aktuelle Kanalkonfig"
          },
          {
            "locale": "pt_BR",
            "description": "remove a configuração atual do canal"
          }
        ],
        "required": false,
        "choices": [ ],
        "options": [ ],
        "autoComplete": false
      },
      {
        "type": "SUB_COMMAND_GROUP",
        "name": "channel_alias",
        "nameLocales": [
          {
            "locale": "de",
            "name": "kanal_alias"
          },
          {
            "locale": "pt_BR",
            "name": "apelidos_do_canal"
          }
        ],
        "description": "add, list or remove channel alias",
        "descriptionLocales": [
          {
            "locale": "de",
            "description": "Kanalkonfig hinzufügen, auflisten und löschen"
          },
          {
            "locale": "pt_BR",
            "description": "adiciona, lista ou remove os apelidos do canal"
          }
        ],
        "required": false,
        "choices": [ ],
        "options": [
          {
            "type": "SUB_COMMAND",
            "name": "save",
            "nameLocales": [
              {
                "locale": "de",
                "name": "speichern"
              },
              {
                "locale": "pt_BR",
                "name": "salvar"
              }
            ],
            "description": "Add a new alias",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Füge einen neuen Alias hinzu"
              },
              {
                "locale": "pt_BR",
                "description": "Adicionar um novo apelido"
              }
            ],
            "required": false,
            "choices": [ ],
            "options": [
              {
                "type": "STRING",
                "name": "name",
                "nameLocales": [
                  {
                    "locale": "pt_BR",
                    "name": "nome"
                  }
                ],
                "description": "The name of the alias (the name in the expression will be replaced with the value)",
                "descriptionLocales": [
                  {
                    "locale": "de",
                    "description": "Der Name des Alias (der Name in den Würfelausdruck wirt mit dem Wert ersetzt)"
                  },
                  {
                    "locale": "pt_BR",
                    "description": "O nome do apelido (o nome da fórmula vai ser substituído com esse valor)"
                  }
                ],
                "required": true,
                "choices": [ ],
                "options": [ ],
                "autoComplete": false
              },
              {
                "type": "STRING",
                "name": "value",
                "nameLocales": [
                  {
                    "locale": "de",
                    "name": "wert"
                  },
                  {
                    "locale": "pt_BR",
                    "name": "valor"
                  }
                ],
                "description": "The value of the alias (the name in the expression will be replaced with the value)",
                "descriptionLocales": [
                  {
                    "locale": "de",
                    "description": "Der Wert des Alias (der Name in den Würfelausdruck wirt mit dem Wert ersetzt)"
                  },
                  {
                    "locale": "pt_BR",
                    "description": "O valor do apelido (o nome pelo qual a expressão vai ser substituída com o valor)"
                  }
                ],
                "required": true,
                "choices": [ ],
                "options": [ ],
                "autoComplete": false
              }
            ],
            "autoComplete": false
          },
          {
            "type": "SUB_COMMAND",
            "name": "delete",
            "nameLocales": [
              {
                "locale": "de",
                "name": "löschen"
              },
              {
                "locale": "pt_BR",
                "name": "deletar"
              }
            ],
            "description": "Delete an alias",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Lösche einen Alias"
              },
              {
                "locale": "pt_BR",
                "description": "Apagar um apelido"
              }
            ],
            "required": false,
            "choices": [ ],
            "options": [
              {
                "type": "STRING",
                "name": "name",
                "nameLocales": [
                  {
                    "locale": "pt_BR",
                    "name": "nome"
                  }
                ],
                "description": "The name of the alias to delete",
                "descriptionLocales": [
                  {
                    "locale": "de",
                    "description": "Der Name des Alias zum Löschen"
                  },
                  {
                    "locale": "pt_BR",
                    "description": "O nome do apelido para apagar"
                  }
                ],
                "required": true,
                "choices": [ ],
                "options": [ ],
                "autoComplete": false
              }
            ],
            "autoComplete": false
          },
          {
            "type": "SUB_COMMAND",
            "name": "list",
            "nameLocales": [
              {
                "locale": "pt_BR",
                "name": "listar"
              }
            ],
            "description": "List all alias",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Liste aller Alias"
              },
              {
                "locale": "pt_BR",
                "description": "Listar todos os apelidos"
              }
            ],
            "required": false,
            "choices": [ ],
            "options": [ ],
            "autoComplete": false
          },
          {
            "type": "SUB_COMMAND",
            "name": "multi_save",
            "nameLocales": [
              {
                "locale": "de",
                "name": "mehrfach_speichern"
              },
              {
                "locale": "pt_BR",
                "name": "salvar_muitos"
              }
            ],
            "description": "Save multiple alias",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Speichere mehrere Alias auf einmal"
              },
              {
                "locale": "pt_BR",
                "description": "Salvar multiplos apelidos"
              }
            ],
            "required": false,
            "choices": [ ],
            "options": [
              {
                "type": "STRING",
                "name": "aliases",
                "nameLocales": [
                  {
                    "locale": "pt_BR",
                    "name": "apelidos"
                  }
                ],
                "description": "Separate alias name and value by `:` and aliases by `;` e.g.: att:2d20;dmg:2d6+3=",
                "descriptionLocales": [
                  {
                    "locale": "de",
                    "description": "Separiere Name und Wert mit `:` und Aliases mit `;` z.B.: Angriff:2d20;Schaden:2d6+3="
                  },
                  {
                    "locale": "pt_BR",
                    "description": "Separe o nome do apelido e o valor com `:` e os apelidos com `;` ex.: atq:2d20;dano:2d6+3="
                  }
                ],
                "required": true,
                "choices": [ ],
                "options": [ ],
                "autoComplete": false
              }
            ],
            "autoComplete": false
          }
        ],
        "autoComplete": false
      },
      {
        "type": "SUB_COMMAND_GROUP",
        "name": "user_channel_alias",
        "nameLocales": [
          {
            "locale": "de",
            "name": "benutzer_kanal_alias"
          },
          {
            "locale": "pt_BR",
            "name": "apelidos_do_usuario"
          }
        ],
        "description": "add, list or remove user channel alias",
        "descriptionLocales": [
          {
            "locale": "de",
            "description": "Benutzer-Kanalkonfig hinzufügen, auflisten und löschen"
          },
          {
            "locale": "pt_BR",
            "description": "adiciona, lista ou remove os apelidos do canal do usuário"
          }
        ],
        "required": false,
        "choices": [ ],
        "options": [
          {
            "type": "SUB_COMMAND",
            "name": "save",
            "nameLocales": [
              {
                "locale": "de",
                "name": "speichern"
              },
              {
                "locale": "pt_BR",
                "name": "salvar"
              }
            ],
            "description": "Add a new alias",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Füge einen neuen Alias hinzu"
              },
              {
                "locale": "pt_BR",
                "description": "Adicionar um novo apelido"
              }
            ],
            "required": false,
            "choices": [ ],
            "options": [
              {
                "type": "STRING",
                "name": "name",
                "nameLocales": [
                  {
                    "locale": "pt_BR",
                    "name": "nome"
                  }
                ],
                "description": "The name of the alias (the name in the expression will be replaced with the value)",
                "descriptionLocales": [
                  {
                    "locale": "de",
                    "description": "Der Name des Alias (der Name in den Würfelausdruck wirt mit dem Wert ersetzt)"
                  },
                  {
                    "locale": "pt_BR",
                    "description": "O nome do apelido (o nome da fórmula vai ser substituído com esse valor)"
                  }
                ],
                "required": true,
                "choices": [ ],
                "options": [ ],
                "autoComplete": false
              },
              {
                "type": "STRING",
                "name": "value",
                "nameLocales": [
                  {
                    "locale": "de",
                    "name": "wert"
                  },
                  {
                    "locale": "pt_BR",
                    "name": "valor"
                  }
                ],
                "description": "The value of the alias (the name in the expression will be replaced with the value)",
                "descriptionLocales": [
                  {
                    "locale": "de",
                    "description": "Der Wert des Alias (der Name in den Würfelausdruck wirt mit dem Wert ersetzt)"
                  },
                  {
                    "locale": "pt_BR",
                    "description": "O valor do apelido (o nome pelo qual a expressão vai ser substituída com o valor)"
                  }
                ],
                "required": true,
                "choices": [ ],
                "options": [ ],
                "autoComplete": false
              }
            ],
            "autoComplete": false
          },
          {
            "type": "SUB_COMMAND",
            "name": "delete",
            "nameLocales": [
              {
                "locale": "de",
                "name": "löschen"
              },
              {
                "locale": "pt_BR",
                "name": "deletar"
              }
            ],
            "description": "Delete an alias",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Lösche einen Alias"
              },
              {
                "locale": "pt_BR",
                "description": "Apagar um apelido"
              }
            ],
            "required": false,
            "choices": [ ],
            "options": [
              {
                "type": "STRING",
                "name": "name",
                "nameLocales": [
                  {
                    "locale": "pt_BR",
                    "name": "nome"
                  }
                ],
                "description": "The name of the alias to delete",
                "descriptionLocales": [
                  {
                    "locale": "de",
                    "description": "Der Name des Alias zum Löschen"
                  },
                  {
                    "locale": "pt_BR",
                    "description": "O nome do apelido para apagar"
                  }
                ],
                "required": true,
                "choices": [ ],
                "options": [ ],
                "autoComplete": false
              }
            ],
            "autoComplete": false
          },
          {
            "type": "SUB_COMMAND",
            "name": "list",
            "nameLocales": [
              {
                "locale": "pt_BR",
                "name": "listar"
              }
            ],
            "description": "List all alias",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Liste aller Alias"
              },
              {
                "locale": "pt_BR",
                "description": "Listar todos os apelidos"
              }
            ],
            "required": false,
            "choices": [ ],
            "options": [ ],
            "autoComplete": false
          },
          {
            "type": "SUB_COMMAND",
            "name": "multi_save",
            "nameLocales": [
              {
                "locale": "de",
                "name": "mehrfach_speichern"
              },
              {
                "locale": "pt_BR",
                "name": "salvar_muitos"
              }
            ],
            "description": "Save multiple alias",
            "descriptionLocales": [
              {
                "locale": "de",
                "description": "Speichere mehrere Alias auf einmal"
              },
              {
                "locale": "pt_BR",
                "description": "Salvar multiplos apelidos"
              }
            ],
            "required": false,
            "choices": [ ],
            "options": [
              {
                "type": "STRING",
                "name": "aliases",
                "nameLocales": [
                  {
                    "locale": "pt_BR",
                    "name": "apelidos"
                  }
                ],
                "description": "Separate alias name and value by `:` and aliases by `;` e.g.: att:2d20;dmg:2d6+3=",
                "descriptionLocales": [
                  {
                    "locale": "de",
                    "description": "Separiere Name und Wert mit `:` und Aliases mit `;` z.B.: Angriff:2d20;Schaden:2d6+3="
                  },
                  {
                    "locale": "pt_BR",
                    "description": "Separe o nome do apelido e o valor com `:` e os apelidos com `;` ex.: atq:2d20;dano:2d6+3="
                  }
                ],
                "required": true,
                "choices": [ ],
                "options": [ ],
                "autoComplete": false
              }
            ],
            "autoComplete": false
          }
        ],
        "autoComplete": false
      }
    ]
  }
]


de.janno.discord.bot.command.channelConfig.ChannelConfigTest.getId=[
  "channel_config"
]